File: civitasjava/src/Civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author info
 */
public class JugadorEspeculador extends Jugador{
    private static int FactorEspeculador = 2;
    private int fianza;
    
    JugadorEspeculador(Jugador j, int f){
        super(j);
        especulador = true;
        fianza = f;
        this.actualizaPropietarioPorConversion();
    }
    
    
    @Override
    protected boolean debeSerEncarcelado(){
        boolean res = false;
        
        if(super.debeSerEncarcelado()){
            if(!puedePagarFianza())
                res = true;
        }
        
        return res;
    }
    
    
    private boolean puedePagarFianza(){
        boolean res = false;
        
        if(saldo >= fianza){
            modificarSaldo(-(fianza));
            res = true;
        }
        
        return res;
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad titulo){
        if(encarcelado)
            return false;
        else{
            if(titulo.getNumHoteles()<getHotelesMax() && 
                    titulo.getNumCasas()>=4)
                return saldo >= titulo.getPrecioEdificar();
            else
                return false;
        }
    }
    
    @Override
    boolean pagaImpuesto(float cantidad){
        if(encarcelado)
            return false;
        else
            return paga(cantidad/FactorEspeculador);
    }
    
    private void actualizaPropietarioPorConversion(){
        for(int i=0; i<super.propiedades.size(); i++){
            propiedades.get(i).setPropietario(this);
        }
    }
    
    @Override
    protected int getCasasMax(){
        return CasasMax*FactorEspeculador;
    }
    
    
    @Override
    protected int getHotelesMax(){
        return HotelesMax*FactorEspeculador;
    }
    
    @Override
    public String toString(){
        return "JugadorEspeculador{ Fianza: " +fianza +" Encarcelado: " 
                +super.encarcelado + ", nombre=" + this.getNombre() + 
                ", numCasillaActual=" + this.getNumCasillaActual() + 
                ", puedeComprar=" + this.getPuedeComprar() + ", saldo=" + 
                saldo + ", propiedades=" + propiedades.toString() +
                ", salvoconducto=" + super.salvoconducto + '}'+"\n";
    }
    
    
    
    
    
    
}

File: civitasruby/lib/jugador_especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class JugadorEspeculador < Jugador
    
    @@FactorEspeculador = 2
    
    def initialize(jugador, fianza)
      jugador_copia(jugador)
      @fianza = fianza
    end
    
    def debe_ser_encarcelado
      res = false
      
      if (super)
        if(!puede_pagar_fianza)
          res = true
        end
      end
    return res
    end
    
    def puede_pagar_fianza
      puedepagar = false
      if(@saldo >= @fianza)
        modificar_saldo(-@fianza)
        puedepagar = true
      end
      return puedepagar
    end
    
    def puedo_edificar_casa(propiedad)
      if(@encarcelado)
        return false
      end
      
      if(propiedad.numCasas < (@@CasasMax * @@FactorEspeculador))
        return @saldo >= propiedad.precioEdificar
      else
        return false
      end
    end
    
    def puedo_edificar_hotel
      if(@encarcelado)
        return false
      end
      if(propiedad.numHoteles< (@@HotelesMax * @@FactorEspeculador) && propiedad.numCasas >=4)
        return @saldo >= propiedad.precioEdificar
      else
        return false
      end
    end
    
    def paga_impuesto(cantidad)
      if(@encarcelado)
        return false
      else
        return paga(cantidad/@@FactorEspeculador)
      end
    end
    
    def get_casas_max
      return (@@CasasMax * @@FactorEspeculador)
    end
    
    def get_hoteles_max
      return (@@HotelesMax * @@FactorEspeciulador)
    end
    
    def to_string
      a= "JugadorEspeculador{fianza= #{@fianza} , nombre= #{@nombre}, saldo= #{@saldo},encarcelado = #{@encarcelado}, numCasillaActual=  #{@numCasillaActual}, puedeComprar= #{@puedeComprar}, salvoconducto= #{@salvoconducto}, propiedades= "
      for i in 0...@propiedades.size
        a = a + @propiedades[i].nombre
      end
      a = a + "}\n"
      return a
    end
    
    def construir_casa(ip)
      res = false
      puedoedificarcasa = false
      if(@encarcelado)
        return res
      else
        existe = existe_la_propiedad(ip)
        if(existe)
          propiedad = @propiedades[ip]
          puedoedificarcasa = self.puedo_edificar_casa(propiedad)
          if(puedoedificarcasa)
            res = propiedad.construir_casa(self)
            if(res)
              Diario.instance.ocurre_evento("El jugador #{@nombre} construye una
                 casa en la propiedad "+ @propiedades[ip].nombre)
            end
          end
        end
      end
      return res
    end
    
    def construir_hotel(ip)
      result = false
      if(@encarcelado)
        return result
      end
      existe=existe_la_propiedad(ip);
      if(existe)
        propiedad = @propiedades[ip]
        puedoEdificarHotel = self.puedo_edificar_hotel(propiedad)
        precio = propiedad.precioEdificar
        if(puedo_gastar(precio) && propiedad.numHoteles < @@HotelesMax && propiedad.numCasas >= @@CasasPorHotel)
          puedoEdificarHotel = true
        end
        if(puedoEdificarHotel)
          result = propiedad.construir_hotel(self)
          propiedad.derruir_casas(@@CasasPorHotel,self)
          Diario.instance.ocurre_evento("El jugador #{@nombre} construye hotel en la propiedad "+ @propiedades[ip].nombre)
      end
    end
    
  end
    
    
  end
end
